icefoganalytics/travel-authorization
# Ytgov Issue 102: Travel Desk Default Flight Dates Do Not Reflect Travel Auth Selection

Fixes https://github.com/ytgov/travel-authorization/issues/102

Relates to:

- https://teams.microsoft.com/l/message/19:0b7c2c4e-e0c8-4708-845d-e0394a499349_56ca4093-1627-46b5-b4a7-e309c41cf283@unq.gbl.spaces/1727196765108?context=%7B%22contextType%22%3A%22chat%22%7D

# Context

**Describe the bug** Once the travel auth. has been approved. The user selects the submit travel desk request. The system added default flights do not fall within the dates selected within the travel auth.

**To Reproduce** Steps to reproduce the behavior:

1. Go to 'My Travel Requests'

2. Click on 'Submit Travel Desk Request'

3. Scroll down to 'Flights section'

4. See error - flight dates and hotels are pulling from the previous submission

**Expected behavior** I would expected that the added flights reflect the dates of the travel auth form.

**Screenshots** ![image](https://private-user-images.githubusercontent.com/89539008/341488714-aac9d2c2-18e1-446b-85c3-ac0055659292.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxOTgwMDAsIm5iZiI6MTcyNzE5NzcwMCwicGF0aCI6Ii84OTUzOTAwOC8zNDE0ODg3MTQtYWFjOWQyYzItMThlMS00NDZiLTg1YzMtYWMwMDU1NjU5MjkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDE3MDgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUzZjg3MDNiZTVmOGVkODBlNDAyMWFmOTBkNDVmMTk0ODkwYmQ4ZTg4ZjgxZGJiYjdhM2YyZWJmMzgwMmM0OWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mkO-29Ev2X_NJQh9PjGg2QEWLIOqwGAWF769b_c6VdU) ![image](https://private-user-images.githubusercontent.com/89539008/341488964-dd69e011-619d-4959-840b-2f586f2808db.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxOTgwMDAsIm5iZiI6MTcyNzE5NzcwMCwicGF0aCI6Ii84OTUzOTAwOC8zNDE0ODg5NjQtZGQ2OWUwMTEtNjE5ZC00OTU5LTg0MGItMmY1ODZmMjgwOGRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDE3MDgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzZWI1NWVmYzU0MzFhNGFkYTM2ZDVmZTc2NjBmMjMyZjFhY2QyNTU2ZmZjZjg0OGIyZjMyMzA4YjRiMzExNmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CVjitq-xXA9k-YIKZR3joLQ1WmHVC9PYuqkTtgyMwQA)

**Desktop (please complete the following information):**

- Browser: Chrome

Update from [external conversation via teams](https://teams.microsoft.com/l/message/19:0b7c2c4e-e0c8-4708-845d-e0394a499349_56ca4093-1627-46b5-b4a7-e309c41cf283@unq.gbl.spaces/1727196765108?context=%7B%22contextType%22%3A%22chat%22%7D) the flight details should be prefilled from the travel segments part of the travel authorization.

The user should have full rights to create/update/delete new travel flights, for now anyway.

# Implementation

1. On travel authorization approval, trigger travel desk creation; now a new service.
2. Travel desk creation now calls the prefill flight requests sub-service, to generate flight information from the parent travel authorization.
3. Lock down flight request travel dates to be during travel authorization timeline.
4. Fix broken production build.
5. Fix a bunch of test flakiness.
6. Split app seeding up by environment, so that the test suite doesn't try to create a bunch of data that it then promptly deletes before starting.
7. Clean up some UI so it works better on mobile.
8. Move DatePicker to /components/common/ and move custom styling out of the component definition. Now "dense" is set during component usage.

# Screenshots

http://localhost:8080/my-travel-requests/5/request/edit?showFlightRequestCreate=true
Flight request date selection is now limited by travel authorization dates.
![image](https://github.com/user-attachments/assets/7e83dc3a-9592-45eb-9ab8-027bd941c4f2)

# Testing Instructions

1. Test the production build via `docker compose build`.
2. Test the production build runs via `docker compose up`
3. Test you can view the production build at http://localhost:3000
4. Run the test suite via `dev test` (or `dev test_api`)
5. Boot the app via `dev up`
6. Log in to the app at http://localhost:8080
7. From the Dashboard page, select "My Travel Requests" from the top dropdown nav.
8. Create a travel authorization via the "+ Travel Authorization" button.
9. Create a travel request between Vancouver and Whitehorse.
10. Pick the travel dates as a short window of 3 days.
11. Generate the estimate.
12. Submit the travel authorization to your email as the supervisor.
13. From the top dropdown nav, select "Manager View".
14. Find your request in the "Pending Approval" card.
15. Approve the request.
16. Return to the "My Travel Requests" page via the top dropdown nav.
17. Find your approved request in the table, and click on the "Submit travel desk request" call-to-action button.
18. Scroll down to the "Flights" section.
19. Note that the flights have been prefilled, and match the travel authorization flight dates.
20. Edit the "Date" field, and note that it is restricted to the travel authorization date range.
21. Update the seat preference, and check that it saves correctly.
